src/Symfony/Component/Validator/Constraints/UniqueValidator.php

```diff
@@ -69,18 +81,46 @@ private function getNormalizer(Unique $unique): callable
         return $unique->normalizer;
     }
 
-    private function reduceElementKeys(array $fields, array $element): array
+    private function reduceElementKeys(array $fields, array|object $element): array
     {
         $output = [];
         foreach ($fields as $field) {
             if (!\is_string($field)) {
                 throw new UnexpectedTypeException($field, 'string');
             }
-            if (\array_key_exists($field, $element)) {
-                $output[$field] = $element[$field];
+
+            $elementAsArray = null;
+            // handle public object property
+            if (\is_object($element) && property_exists($element, $field)) {
+                $elementAsArray = (array) $element;
+            } elseif (\is_array($element)) {
+                $elementAsArray = $element;
+            }
+
+            if ($elementAsArray && \array_key_exists($field, $elementAsArray)) {
+                $output[$field] = $elementAsArray[$field];
+                continue;
+            }
+
+            try {
+                $output[$field] = $this->getPropertyAccessor()->getValue($element, $field);
+            } catch (AccessException) {
+                // fields are optional
             }
         }
 
         return $output;
     }
+
+    private function getPropertyAccessor(): PropertyAccessor
+    {
+        if (null === $this->propertyAccessor) {
+            if (!class_exists(PropertyAccess::class)) {
+                throw new LogicException('Property path requires symfony/property-access package to be installed.');
```

Comment:
It would be great for DX if we could write the message as follow:
```suggestion
                throw new LogicException('Property path requires symfony/property-access package to be installed. Try running "composer require symfony/property-access".');
```
Additionally, it would be consistent with the others messages as seen here: https://github.com/symfony/symfony/search?q=%22composer+require%22